Compatibility with QuickCheck-2.12

Now that `QuickCheck-2.12` has been released, it would be handy to have a version of `test-framework-quickcheck2` which supported it. Currently, trying to bump the upper version bounds on `QuickCheck` will lead to a build failure:

```
[1 of 1] Compiling Test.Framework.Providers.QuickCheck2 ( Test/Framework/Providers/QuickCheck2.hs, /home/rgscott/Documents/Hacking/Haskell/test-framework/dist-newstyle/build/x86_64-linux/ghc-8.4.3/test-framework-quickcheck2-0.3.0.4/build/Test/Framework/Providers/QuickCheck2.o )

Test/Framework/Providers/QuickCheck2.hs:131:23: error:
    Not in scope: data constructor ‘InsufficientCoverage’
    |
131 |     toPropertyStatus (InsufficientCoverage _ _ _)              = PropertyInsufficientCoverage
    |                       ^^^^^^^^^^^^^^^^^^^^
```

This can be worked around easily enough:

```diff
diff --git a/quickcheck2/Test/Framework/Providers/QuickCheck2.hs b/quickcheck2/Test/Framework/Provi
index b61b808..6109400 100644
--- a/quickcheck2/Test/Framework/Providers/QuickCheck2.hs
+++ b/quickcheck2/Test/Framework/Providers/QuickCheck2.hs
@@ -127,6 +127,6 @@ runProperty topts testable = do
     toPropertyStatus (GaveUp {})                               = PropertyArgumentsExhausted
     toPropertyStatus (Failure { reason = rsn, output = otpt }) = PropertyFalsifiable rsn otpt
     toPropertyStatus (NoExpectedFailure {})                    = PropertyNoExpectedFailure
-#if MIN_VERSION_QuickCheck(2,8,0)
+#if MIN_VERSION_QuickCheck(2,8,0) && !MIN_VERSION_QuickCheck(2,12,0)
     toPropertyStatus (InsufficientCoverage _ _ _)              = PropertyInsufficientCoverage
 #endif
```

But if we do this, then `PropertyInsufficientCoverage` becomes unused when `test-framework-quickcheck2` is built against `QuickCheck-2.12`. Should we CPP out `PropertyInsufficientCoverage` as well with `QuickCheck-2.12` (since it's already conditionally defined on `MIN_VERSION_QuickCheck(2,8,0)` anyway)?

--- a/core/test-framework.cabal
+++ b/core/test-framework.cabal
@@ -104,7 +104,7 @@ Test-Suite test-framework-tests
 
         -- Buildable:              False
         Build-Depends:          HUnit          >= 1.2
-                              , QuickCheck     >= 2.3 && < 2.12
+                              , QuickCheck     >= 2.3 && < 2.13
                               , base           >= 4.3
                               , random         >= 1.0
                               , containers     >= 0.1

--- a/example/test-framework-example.cabal
+++ b/example/test-framework-example.cabal
@@ -23,6 +23,6 @@ Executable test-framework-example
 
         Build-Depends:          base                       >= 3   && < 5
                               , HUnit                      >= 1.2 && < 1.7
-                              , QuickCheck                 >= 2.4 && < 2.12
+                              , QuickCheck                 >= 2.4 && < 2.13
 
         Ghc-Options:            -threaded

--- a/quickcheck2/Test/Framework/Providers/QuickCheck2.hs
+++ b/quickcheck2/Test/Framework/Providers/QuickCheck2.hs
@@ -49,7 +49,7 @@ data PropertyStatus = PropertyOK                        -- ^ The property is tru
                     | PropertyFalsifiable String String -- ^ The property was not true. The strings are the reason and the output.
                     | PropertyNoExpectedFailure         -- ^ We expected that a property would fail but it didn't
                     | PropertyTimedOut                  -- ^ The property timed out during execution
-#if MIN_VERSION_QuickCheck(2,8,0)
+#if MIN_VERSION_QuickCheck(2,8,0) && !MIN_VERSION_QuickCheck(2,12,0)
                     | PropertyInsufficientCoverage      -- ^ The tests passed but a use of 'cover' had insufficient coverage.
 #endif
 
@@ -61,7 +61,7 @@ instance Show PropertyResult where
             PropertyFalsifiable _rsn otpt -> otpt ++ "(used seed " ++ show used_seed ++ ")"
             PropertyNoExpectedFailure     -> "No expected failure with seed " ++ show used_seed ++ ", after " ++ tests_run_str ++ " tests"
             PropertyTimedOut              -> "Timed out after " ++ tests_run_str ++ " tests"
-#if MIN_VERSION_QuickCheck(2,8,0)
+#if MIN_VERSION_QuickCheck(2,8,0) && !MIN_VERSION_QuickCheck(2,12,0)
             PropertyInsufficientCoverage  -> "Insufficient coverage after " ++ tests_run_str ++ " tests"
 #endif
       where
@@ -127,6 +127,6 @@ runProperty topts testable = do
     toPropertyStatus (GaveUp {})                               = PropertyArgumentsExhausted
     toPropertyStatus (Failure { reason = rsn, output = otpt }) = PropertyFalsifiable rsn otpt
     toPropertyStatus (NoExpectedFailure {})                    = PropertyNoExpectedFailure
-#if MIN_VERSION_QuickCheck(2,8,0)
+#if MIN_VERSION_QuickCheck(2,8,0) && !MIN_VERSION_QuickCheck(2,12,0)
     toPropertyStatus (InsufficientCoverage _ _ _)              = PropertyInsufficientCoverage
 #endif

--- a/quickcheck2/test-framework-quickcheck2.cabal
+++ b/quickcheck2/test-framework-quickcheck2.cabal
@@ -26,7 +26,7 @@ Library
         Exposed-Modules:        Test.Framework.Providers.QuickCheck2
 
         Build-Depends:          test-framework        == 0.8.*
-                              , QuickCheck            >= 2.4    && < 2.12
+                              , QuickCheck            >= 2.4    && < 2.13
                               , base                  >= 4.3    && < 5
                               , extensible-exceptions >= 0.1.1  && < 0.2.0
                               , random                >= 1      && < 1.2

